### PR TITLE
U4-11167 Fixes rare syncTree error

### PR DESCRIPTION
**Pull request by dawoe**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have linked this PR to an issue on the tracker at http://issues.umbraco.org/issue/U4-11167

### Description

It sometimes occurs when you do a hard refresh of your browser that the syncTree method is called in the navigation service, before the umb-tree directive has finished with setting up external events in this method : https://github.com/umbraco/Umbraco-CMS/blob/dev-v7/src/Umbraco.Web.UI.Client/src/common/directives/components/tree/umbtree.directive.js#L84

I wrapped in a timeout so we and could not reproduce it anymore

### CODE REVIEW

File: src/Umbraco.Web.UI.Client/src/common/services/navigation.service.js [hunk 1]
```diff
@@ -288,9 +288,12 @@ function navigationService($rootScope, $routeParams, $log, $location, $q, $timeo
                 throw "args.tree cannot be null";
             }
 
-            if (mainTreeEventHandler) {
-                //returns a promise
-                return mainTreeEventHandler.syncTree(args);
+            if (mainTreeEventHandler) {                
+                //returns a promise, but wrap it in a time out,
+                // on some occasions the syncTree method is not defined yet in the umb - tree directive(U4 - 11167)
+                $timeout(function () {
```
Comment (Shazwazza) [inline src/Umbraco.Web.UI.Client/src/common/services/navigation.service.js]: this changes how the logic flows. Before we are returning the result of `mainTreeEventHandler.syncTree(args);` but now it is returning nothing. Returning from within the $timeout is not the same thing.

We are already checking if `mainTreeEventHandler` exists, why not also check for `mainTreeEventHandler.syncTree`, perhaps this check is all that is needed

Comment (dawoe) [reply src/Umbraco.Web.UI.Client/src/common/services/navigation.service.js]: Hi Shannon,

That was my first solution. But the side effect was that the tree never selected the current node as active. If this problem occurs in 4th level content item, the content tree remains collapsed.

That's why I used the timeout. Maybe only execute the timeout when syncTree is not defined ?

Comment (dawoe) [reply src/Umbraco.Web.UI.Client/src/common/services/navigation.service.js]: I updated it to only use a timeout when the syncTree method is not defined

File: src/Umbraco.Web.UI.Client/src/common/services/navigation.service.js [hunk 2]
```diff
@@ -288,9 +288,18 @@ function navigationService($rootScope, $routeParams, $log, $location, $q, $timeo
...
-                return mainTreeEventHandler.syncTree(args);
+            if (mainTreeEventHandler) {                
+               
+                if (mainTreeEventHandler.syncTree) {
+                    //returns a promise,
+                    return mainTreeEventHandler.syncTree(args);
+                } else {
+                    //returns a promise, but wrap it in a time out,
+                    // on some occasions the syncTree method is not defined yet in the umb - tree directive(U4 - 11167)
+                    $timeout(function () {
```
Comment (Shazwazza) [inline src/Umbraco.Web.UI.Client/src/common/services/navigation.service.js]: This is still not actually returning a promise :) this is executing a function after a timeout and that inner function is returning a value to the $timeout method, not to the caller of `syncTree`

I'm wondering if we can instead:

* Error check that `mainTreeEventHandler.syncTree` is defined and if not throw an exception - if we are expecting it to always be set
* eagerly ensure that `mainTreeEventHandler.syncTree` is set before `navigationService.syncTree` is ever executed

Comment (dawoe) [reply src/Umbraco.Web.UI.Client/src/common/services/navigation.service.js]: Hi Shannon,

Maybe just suppress the error for now when syncTree is not defined. This bug has been present for a longer time. I could trace it back to 7.8.0, but I did not try earlier versions.

It also only happens when special conditions are met...eg. doing a hard refresh and with empty cache. And then not even every time.

Dave

**Pull request merged**